File: libcudf-sys/src/stream.cpp
#include "stream.h"

#include <stdexcept>

namespace libcudf_bridge {
    namespace {
        constexpr uint32_t kCudaStreamFlagSyncDefault = 0;
        constexpr uint32_t kCudaStreamFlagNonBlocking = 1;

        static_assert(
            static_cast<uint32_t>(rmm::cuda_stream::flags::sync_default) ==
            kCudaStreamFlagSyncDefault);
        static_assert(
            static_cast<uint32_t>(rmm::cuda_stream::flags::non_blocking) ==
            kCudaStreamFlagNonBlocking);

        [[nodiscard]] rmm::cuda_stream::flags to_rmm_flags(const uint32_t flags) {
            return static_cast<rmm::cuda_stream::flags>(flags);
        }
    } // namespace

    CudaStreamView::CudaStreamView(rmm::cuda_stream_view stream) : inner(stream) {}

    CudaStreamView::~CudaStreamView() = default;

    bool CudaStreamView::is_default() const {
        return inner.is_default();
    }

    bool CudaStreamView::is_per_thread_default() const {
        return inner.is_per_thread_default();
    }

    void CudaStreamView::synchronize() const {
        inner.synchronize();
    }

    /// By default, create a stream that synchronizes with the default stream
    /// (ie. this uses [`rmm::cuda_stream::flags::sync_default`]).
    CudaStream::CudaStream() : inner(std::make_unique<rmm::cuda_stream>()) {}

    CudaStream::CudaStream(const uint32_t flags)
        : inner(std::make_unique<rmm::cuda_stream>(to_rmm_flags(flags))) {}

    CudaStream::~CudaStream() = default;

    // A wrapper is valid as long as it still owns an underlying RMM stream.
    bool CudaStream::is_valid() const {
        return inner && inner->is_valid();
    }

    void CudaStream::synchronize() const {
        if (!inner) {
            throw std::runtime_error("Cannot synchronize null CUDA stream");
        }
        inner->synchronize();
    }

    rmm::cuda_stream_view CudaStream::view() const {
        // In case `inner` gets moved by assigning one `CudaStream` to another.
        if (!inner) {
            throw std::runtime_error("Cannot get view of null CUDA stream");
        }
        return inner->view();
    }

    std::unique_ptr<CudaStream> cuda_stream_create() {
        return std::make_unique<CudaStream>();
    }

    std::unique_ptr<CudaStream> cuda_stream_create_with_flags(const uint32_t flags) {
        return std::make_unique<CudaStream>(flags);
    }

    std::unique_ptr<CudaStreamView> cuda_stream_view(const CudaStream& stream) {
        return std::make_unique<CudaStreamView>(stream.view());
    }

    std::unique_ptr<CudaStreamView> get_default_stream() {
        return std::make_unique<CudaStreamView>(cudf::get_default_stream());
    }

    bool is_ptds_enabled() {
        return cudf::is_ptds_enabled();
    }
} // namespace libcudf_bridge

File: src/stream.rs
use cxx::UniquePtr;

/// Stream creation flags for CUDA stream-backed cuDF execution.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CuDFStreamFlags {
    /// Create a stream that synchronizes with the default stream.
    SyncDefault,
    /// Create a non-blocking stream that does not synchronize with the default stream.
    NonBlocking,
}

const CUDA_STREAM_FLAG_SYNC_DEFAULT: u32 = 0;
const CUDA_STREAM_FLAG_NON_BLOCKING: u32 = 1;

impl From<CuDFStreamFlags> for u32 {
    fn from(value: CuDFStreamFlags) -> Self {
        match value {
            CuDFStreamFlags::SyncDefault => CUDA_STREAM_FLAG_SYNC_DEFAULT,
            CuDFStreamFlags::NonBlocking => CUDA_STREAM_FLAG_NON_BLOCKING,
        }
    }
}

/// Owning Rust wrapper for a CUDA stream used by cuDF operations.
///
/// This type owns an opaque C++ `rmm::cuda_stream`. Dropping `CuDFStream`
/// destroys that underlying stream.
///
/// The handle may be shared across host threads. Work enqueued onto the same
/// stream executes in order; sharing the handle does not by itself synchronize
/// access to GPU memory used by those operations.
pub struct CuDFStream {
    inner: UniquePtr<libcudf_sys::ffi::CudaStream>,
}

impl CuDFStream {
    /// Create a stream using the sync-default creation flag.
    pub fn new() -> Self {
        Self {
            inner: libcudf_sys::ffi::cuda_stream_create(),
        }
    }

    /// Create a stream with explicit creation flags.
    pub fn with_flags(flags: CuDFStreamFlags) -> Self {
        Self {
            inner: libcudf_sys::ffi::cuda_stream_create_with_flags(flags.into()),
        }
    }

    /// Get a reference to the underlying FFI stream handle.
    pub(crate) fn inner(&self) -> &libcudf_sys::ffi::CudaStream {
        self.inner.as_ref().expect("CudaStream should not be null")
    }
}

impl Default for CuDFStream {
    fn default() -> Self {
        Self::new()
    }
}
